### PR TITLE
Rich-text support for comments and response text #5830

### LINKED ISSUE #5830: Rich-text support for comments and response text
$subj

### PR DESCRIPTION
**Pull request by belyabl9**
Fixes #5830

### CODE REVIEW

**Comment by belyabl9** [comment]
Ready to review

**Comment by wkurniawan07** [comment]
@belyabl9 correct me if I'm wrong, but this PR includes support for response texts as well?

**Comment by belyabl9** [comment]
@wkurniawan07 , you're right. I'll change the issue and PR name

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorCommentsPageUiTest.java [hunk 1]
```diff
@@ -179,6 +180,7 @@ private void testPanelsCollapseExpand() {
     private void testActions() throws Exception {
         ______TS("action: edit student comment");
         commentsPage.clickStudentCommentEditForRow(1);
+        ThreadHelper.waitFor(1000);
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorCommentsPageUiTest.java]: Fixed time waits like this are not preferred in general. What is this line waiting for?

File: src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java [hunk 2]
```diff
@@ -43,7 +43,8 @@ public void fillResponseTextBox(int qnNumber, int responseNumber, String text) {
         WebElement element = browser.driver.findElement(
                 By.name(Const.ParamsNames.FEEDBACK_RESPONSE_TEXT + "-" + qnNumber + "-" + responseNumber));
         click(element);
-        fillTextBox(element, text);
+        element.clear();
+        element.sendKeys(text);
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java]: Why does the previous `fillTextBox(...)` not work?

**Comment by wkurniawan07** [comment]
Two things:
- Using `ThreadHelper.waitFor(...)` is in general discouraged. The preferred way is to add explicit wait for the expected condition. There are many examples of this in `AppPage.java`.
- If there are `element.click()`, let's change it to `click(element)` so that Selenium can fully support it, unless there are good reasons to retain the first one?

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorCommentsPageUiTest.java (hunk 1 above)
Comment (belyabl9) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorCommentsPageUiTest.java]: It waits for TinyMCE initialization, because lazy initialization is used i.e. only after showing edit row

File: src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java (hunk 2 above)
Comment (belyabl9) [inline src/test/java/teammates/test/pageobjects/FeedbackSubmitPage.java]: It should be enough. Initially, I changed it to fillRichTextEditor() but such approach doesn't handle symbols counter, so I came back. I'll use fillTextBox(), it seems to be the same

**Comment by belyabl9** [comment]
Ready to review

File: src/main/webapp/WEB-INF/tags/shared/commentRow.tag [hunk 3]
```diff
@@ -262,12 +262,12 @@
...
                 <div class="form-group">
-                    <%-- Do not add whitespace between the opening and closing tags--%>
-                    <textarea class="form-control"
-                              rows="3"
-                              placeholder="Your comment about this student"
-                              name="<%= Const.ParamsNames.COMMENT_TEXT %>"
-                              id="commentText${commentIndex}">${comment.commentText}</textarea>
+                    <div id="rich-text-toolbar-comment-container-${commentIndex}"></div>
+                    <div class="panel panel-default panel-body"
+                         id="commentText${commentIndex}">
```
Comment (wkurniawan07) [inline src/main/webapp/WEB-INF/tags/shared/commentRow.tag]: This should be fine on the previous line.

File: src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag [hunk 4]
```diff
@@ -189,11 +189,13 @@
...
     <div class="form-group">
-        <textarea class="form-control"
-                  rows="3"
-                  placeholder="Your comment about this response"
-                  name="<%= Const.ParamsNames.FEEDBACK_RESPONSE_COMMENT_TEXT %>"
-                  id="${textAreaId}-${divId}">${frc.commentText}</textarea>
+        <div id="rich-text-toolbar-comment-container-${divId}"></div>
+        <div class="panel panel-default panel-body"
+             rows="3"
+             id="${textAreaId}-${divId}">
```
Comment (wkurniawan07) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag]: Here as well, in fact starting from `rows="3"`.

File: src/main/webapp/js/common.js [hunk 5]
```diff
@@ -884,6 +884,16 @@ function hideSingleCollapse(e) {
     $(heading).find('a.btn').hide();
 }
 
+function destroyEditor(id) {
```
Comment (wkurniawan07) [inline src/main/webapp/js/common.js]: While we're at it, let's add a header comment for this function.

File: src/test/java/teammates/test/pageobjects/InstructorCommentsPage.java [hunk 6]
```diff
@@ -174,11 +175,11 @@ public void fillTextareaToEditResponseComment(int sessionIdx, int questionIdx, i
...
                                                   int commentIdx, String text) {
+        String editorId = "responsecommenttext-" + sessionIdx + "-" + questionIdx + "-" + responseIdx + "-" + commentIdx;
         WebElement textarea = browser.driver.findElement(
-                By.id("responsecommenttext-" + sessionIdx + "-" + questionIdx + "-" + responseIdx + "-" + commentIdx));
-        click(textarea);
-        textarea.clear();
-        textarea.sendKeys(text);
+                By.id(editorId));
+        waitForRichTextEditorToLoad(editorId);
+        fillTextBox(textarea, text);
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/InstructorCommentsPage.java]: Why is this not `fillRichTextEditor`?

File: src/test/java/teammates/test/pageobjects/InstructorCommentsPage.java [hunk 7]
```diff
@@ -174,11 +175,11 @@ public void fillTextareaToEditResponseComment(int sessionIdx, int questionIdx, i
     
     public void fillTextareaToEditResponseComment(int sessionIdx, int questionIdx, int responseIdx,
                                                   int commentIdx, String text) {
+        String editorId = "responsecommenttext-" + sessionIdx + "-" + questionIdx + "-" + responseIdx + "-" + commentIdx;
         WebElement textarea = browser.driver.findElement(
-                By.id("responsecommenttext-" + sessionIdx + "-" + questionIdx + "-" + responseIdx + "-" + commentIdx));
-        click(textarea);
-        textarea.clear();
-        textarea.sendKeys(text);
+                By.id(editorId));
```
Comment (wkurniawan07) [inline src/test/java/teammates/test/pageobjects/InstructorCommentsPage.java]: Again this one should be fine on the previous line.

File: src/test/resources/pages/instructorCommentsPageAddFrc.html [hunk 8]
```diff
@@ -1809,7 +1839,7 @@ <h4 class="col-sm-9 text-color-primary">
                             </a>
                           </div>
                           <div id="plainCommentText-1-1-1-1" style="margin-left: 15px;">
-                            added response comment
+                            Instructor 1 comment to student 1 self feedback
```
Comment (wkurniawan07) [inline src/test/resources/pages/instructorCommentsPageAddFrc.html]: Any idea why the previous response comment is removed?

File: src/test/resources/pages/instructorCommentsPageDeleteFrc.html [hunk 9]
```diff
@@ -1744,9 +1778,9 @@ <h4 class="col-sm-9 text-color-primary">
             <div class="panel panel-info feedback-question-panel">
               <div class="panel-heading">
                 <b>
-                  Question 1
+                  Question 2
```
Comment (wkurniawan07) [inline src/test/resources/pages/instructorCommentsPageDeleteFrc.html]: The first question seems to get removed. I suspect it's an off-by-one error somewhere?

**Comment by wkurniawan07** [comment]
Added some comments, the most important one being there are functions that might have been broken. Also there are merge conflicts to be resolved.

**Comment by belyabl9** [comment]
@wkurniawan07 , I'll check it in more detail, but I think that the cause of removed questions is in InstructorStudentCommentEditAction.java. I removed assertion for deletion with empty comment text, can't get why it was needed if comment has identifier

File: src/test/resources/pages/instructorSearchPageSearchStudentComments.html [hunk 10]
```diff
@@ -135,7 +135,7 @@
                   </b>
                   [Tue, 01 May 2012, 08:49 AM UTC]
                 </span>
-                <a class="btn btn-default btn-xs icon-button pull-right" data-original-title="Edit comment in the Comments page" data-placement="top" data-toggle="tooltip" href="/page/instructorCommentsPage?user=searchUI.idOfInstructor1OfCourse1&courseid=searchUI.idOfTypicalCourse1#${comment.id}" style="display:none;" target="_blank" title="" type="button">
+                <a class="btn btn-default btn-xs icon-button pull-right" data-original-title="Edit comment in the Comments page" data-placement="top" data-toggle="tooltip" href="/page/instructorCommentsPage?user=searchUI.idOfInstructor1OfCourse1&courseid=searchUI.idOfTypicalCourse1#${comment.id}" style="" target="_blank" title="" type="button">
```
Comment (wkurniawan07) [inline src/test/resources/pages/instructorSearchPageSearchStudentComments.html]: This change should not occur. I have tried this branch on my own local environment.

File: src/test/resources/pages/instructorSearchPageSearchComments.html [hunk 11]
```diff
@@ -135,7 +135,7 @@
                   </b>
                   [Tue, 01 May 2012, 08:49 AM UTC]
                 </span>
-                <a class="btn btn-default btn-xs icon-button pull-right" data-original-title="Edit comment in the Comments page" data-placement="top" data-toggle="tooltip" href="/page/instructorCommentsPage?user=searchUI.idOfInstructor1OfCourse1&courseid=searchUI.idOfTypicalCourse1#${comment.id}" style="display:none;" target="_blank" title="" type="button">
+                <a class="btn btn-default btn-xs icon-button pull-right" data-original-title="Edit comment in the Comments page" data-placement="top" data-toggle="tooltip" href="/page/instructorCommentsPage?user=searchUI.idOfInstructor1OfCourse1&courseid=searchUI.idOfTypicalCourse1#${comment.id}" style="" target="_blank" title="" type="button">
```
Comment (wkurniawan07) [inline src/test/resources/pages/instructorSearchPageSearchComments.html]: Similarly here.

Comment (belyabl9) [inline src/test/resources/pages/instructorSearchPageSearchComments.html]: I have a problem with it. Actual result is variable

Comment (wkurniawan07) [inline src/test/resources/pages/instructorSearchPageSearchComments.html]: I can't replicate the problem, it's consistently returning the correct one (with `display:none;`) for me. Same with your other PR.

Comment (belyabl9) [inline src/test/resources/pages/instructorSearchPageSearchComments.html]: It's strange. What about FF version, as I remember it was updated for testing. I am still on previous one. May it be the reason ?

Comment (wkurniawan07) [inline src/test/resources/pages/instructorSearchPageSearchComments.html]: Unlikely, because I just tried with the old 38.0.5 and it still consistently returns the correct one.
Anyway, let's move forward with this PR: undo these changes, they're preventing a passing build in Travis.

File: src/test/resources/pages/instructorCommentsPageAddSc.html [hunk 12]
```diff
@@ -227,7 +227,7 @@ <h4 class="col-sm-9 text-color-primary">
                       <b>
                         all students in this course
                       </b>
-                      [Fri, 01 Jun 2012, 07:20 PM UTC] (last edited by comments.instructor1@course1.tmt at ${datetime.now})
+                      [Fri, 01 Jun 2012, 07:20 PM UTC] (last edited by comments.instructor1@course1.tmt at Thu, 28 Jul 2016, 09:54 PM UTC)
```
Comment (wkurniawan07) [inline src/test/resources/pages/instructorCommentsPageAddSc.html]: You need to run the tests with `-Duser.timezone=UTC` VM args, otherwise there are certain times in the day where this kind of unwanted replacement happens.

**Comment by belyabl9** [comment]
Ready to review

**Comment by wkurniawan07** [comment]
@belyabl9 I've identified what caused the missing comment in the `instructorCommentsPageEditFrc.html`. It doesn't look like anything to do with Firefox cache or alike. As it happens, this action on line 230 of `InstructorCommentsPageUiTest`:
```java
commentsPage.saveResponseComment(1, 1, 1, 1);
```
actually did not execute and returns "Comment cannot be empty" for some reason. Can you fix that?

**Comment by belyabl9** [comment]
@wkurniawan07 , I've fixed that issue. Ready to review

**Comment by belyabl9** [comment]
Please, review

File: src/main/java/teammates/common/datatransfer/CommentAttributes.java [hunk 13]
```diff
@@ -48,7 +48,7 @@ public CommentAttributes(String courseId, String giverEmail, CommentParticipantT
         this.giverEmail = giverEmail;
         this.recipientType = recipientType == null ? CommentParticipantType.PERSON : recipientType;
         this.recipients = recipients;
-        this.commentText = commentText;
+        this.commentText = commentText == null ? null : new Text(Sanitizer.sanitizeForRichText(commentText.getValue()));
```
Comment (wkurniawan07) [inline src/main/java/teammates/common/datatransfer/CommentAttributes.java]: Just thinking, is it possible to make a `sanitizeForRichText` method that takes in `Text` parameter? It can be reused here and potentially many other places.

File: src/main/webapp/WEB-INF/tags/shared/commentRow.tag [hunk 14]
```diff
@@ -262,12 +262,11 @@
...
-                    <textarea class="form-control"
-                              rows="3"
-                              placeholder="Your comment about this student"
-                              name="<%= Const.ParamsNames.COMMENT_TEXT %>"
-                              id="commentText${commentIndex}">${comment.commentText}</textarea>
+                    <div id="rich-text-toolbar-comment-container-${commentIndex}"></div>
+                    <div class="panel panel-default panel-body" id="commentText${commentIndex}">
+                         ${comment.commentText}
+                    </div>
+                    <input type="hidden" name="commenttext">
```
Comment (wkurniawan07) [inline src/main/webapp/WEB-INF/tags/shared/commentRow.tag]: `commenttext` should be taken from a constant in `Const.ParamsNames`, right?

File: src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag [hunk 15]
```diff
@@ -189,11 +189,11 @@
...
-        <textarea class="form-control"
-                  rows="3"
-                  placeholder="Your comment about this response"
-                  name="<%= Const.ParamsNames.FEEDBACK_RESPONSE_COMMENT_TEXT %>"
-                  id="${textAreaId}-${divId}">${frc.commentText}</textarea>
+        <div id="rich-text-toolbar-comment-container-${divId}"></div>
+        <div class="panel panel-default panel-body" id="${textAreaId}-${divId}">
+            ${frc.commentText}
+        </div>
+        <input type="hidden" name="responsecommenttext">
```
Comment (wkurniawan07) [inline src/main/webapp/WEB-INF/tags/shared/feedbackResponseCommentForm.tag]: Similarly for `responsecommenttext`.

File: src/main/webapp/js/common.js [hunk 16]
```diff
@@ -887,6 +887,19 @@ function hideSingleCollapse(e) {
 }
 
 /**
+ * Destroys an instance of TinyMCE rich-text editor.
+ */
+function destroyEditor(id) {
```
Comment (wkurniawan07) [inline src/main/webapp/js/common.js]: Move this to `richTextEditor.js`?

File: src/main/webapp/js/omniComment.js [hunk 17]
```diff
@@ -402,6 +410,16 @@ function enableComment(commentIdx) {
...
     $("textarea[id='commentText" + commentIdx + "']").val($('#plainCommentText' + commentIdx).text());
+
+    if (typeof richTextEditorBuilder !== 'undefined') {
+        /* eslint-disable camelcase */ // The property names are determined by external library (tinymce)
+        richTextEditorBuilder.initEditor('#commentText' + commentIdx, {
+            inline: true,
+            fixed_toolbar_container: '#rich-text-toolbar-comment-container-' + commentIdx
+        });
+        /* eslint-enable camelcase */
+    }
```
Comment (wkurniawan07) [inline src/main/webapp/js/omniComment.js]: Actually this routine looks repeated many times as well. Can they be extracted to `richTextEditor.js`?

**Comment by wkurniawan07** [comment]
Added a few comments to address, some possibly not in this PR.

File: src/main/webapp/js/omniComment.js (hunk 17 above)
Comment (belyabl9) [inline src/main/webapp/js/omniComment.js]: I think it can't. It's already a routine in richTextEditor.js and it has variable params like 'fixed_toolbar_container' and selector

File: src/main/java/teammates/common/datatransfer/CommentAttributes.java (hunk 13 above)
Comment (belyabl9) [inline src/main/java/teammates/common/datatransfer/CommentAttributes.java]: As it's used in other places unrelated to this issue, it'd make sense to do it in a separate issue

**Comment by belyabl9** [comment]
Ready to review

**Comment by belyabl9** [comment]
Please, review

File: src/main/webapp/js/contextualcomments.js [hunk 18]
```diff
@@ -4,16 +4,25 @@ $(document).ready(function() {
...
+        /* eslint-disable camelcase */ // The property names are determined by external library (tinymce)
+        richTextEditorBuilder.initEditor('#commenttext', {
+            inline: true,
+            fixed_toolbar_container: '#rich-text-toolbar-comment-container'
+        });
+        /* eslint-enable camelcase */
+    }
+
     $('form[name="form_commentadd"]').submit(function() {
+        tinymce.get('commenttext').save();
```
Comment (wkurniawan07) [inline src/main/webapp/js/contextualcomments.js]: What's the difference between `tinymce` and `tinyMCE`?

Comment (belyabl9) [inline src/main/webapp/js/contextualcomments.js]: As I understand, both variants are possible, like aliases

File: src/test/resources/pages/InstructorCourseDetailsEmptyCourse.html [hunk 19]
```diff
@@ -1,5 +1,8 @@
 <html>
 <head>
+<style id="mceDefaultStyles" type="text/css">
+  .mce-content-body div.mce-resizehandle {position: absolute;border: 1px solid black;background: #FFF;width: 5px;height: 5px;z-index: 10000}.mce-content-body .mce-resizehandle:hover {background: #000}.mce-content-body img[data-mce-selected], hr[data-mce-selected] {outline: 1px solid black;resize: none}.mce-content-body .mce-clonedresizable {position: absolute;opacity: .5;filter: alpha(opacity=50);z-index: 10000}.mce-content-body .mce-resize-helper {background: #555;background: rgba(0,0,0,0.75);border-radius: 3px;border: 1px;color: white;display: none;font-family: sans-serif;font-size: 12px;white-space: nowrap;line-height: 14px;margin: 5px 10px;padding: 5px;position: absolute;z-index: 10001} img:-moz-broken {-moz-force-broken-image-icon:1;min-width:24px;min-height:24px}
+</style>
```
Comment (wkurniawan07) [inline src/test/resources/pages/InstructorCourseDetailsEmptyCourse.html]: You need to regenerate this HTML file again, a placeholder `${tinymce.style}` is expected here due to #5934. Similarly for other files that have these 3 lines.

**Comment by wkurniawan07** [comment]
Changes are good, just that one minor thing needs to be done on some HTML test files.

**Comment by belyabl9** [comment]
@wkurniawan07 , please, review
Travis results are a little bit strange

**Comment by wkurniawan07** [comment]
@belyabl9 apologies, there's little I can do here until Travis decides to cooperate with us again.

**Comment by wkurniawan07** [comment]
@belyabl9 can you sync this branch with the current `master`? I need to see if the problem is still there.

**Comment by belyabl9** [comment]
@wkurniawan07 , done. As I see, the issue persists

**Comment by damithc** [comment]
Can this be wrapped up soon?

**Comment by belyabl9** [comment]
@wkurniawan07 promised (in slack) to have a look at this strange Travis error, so I'll proceed after his comment

**Comment by wkurniawan07** [comment]
I'm at a loss for the behaviour myself. Since we don't really have the luxury of time, I'll be circumventing this problem by ignoring the guilty placeholder. Putting this on hold for a very short while, will update later.

**Comment by wkurniawan07** [comment]
@belyabl9 this can be continued now

File: src/test/resources/pages/InstructorCourseDetailsEmptyCourse.html (hunk 19 above)
Comment (wkurniawan07) [inline src/test/resources/pages/InstructorCourseDetailsEmptyCourse.html]: These lines will be ignored anyway, so can you remove them? Similarly for the other files.

**Comment by belyabl9** [comment]
Ready to review

**Comment by wkurniawan07** [comment]
Everything looks good except for that possible off-by-one error (dated more than a month back). If it's indeed caused by the removed assertion, can you revert it for now? It's better not to have unexpected behavior changes at this point of time. If it's not the cause, then it's something that needs to be fixed.

**Comment by damithc** [comment]
@belyabl9 for your followup please.

**Comment by belyabl9** [comment]
I'll try to check it this week. My current schedule is more tight than in summer

**Comment by wkurniawan07** [comment]
@belyabl9 any updates here?

**Comment by belyabl9** [comment]
@wkurniawan07 , I'll continue to reproduce this issue today, By the way, as I see from the comments history I've already fixed that.

**Comment by wkurniawan07** [comment]
@belyabl9 it has not been fixed yet. If you are referring to [this comment](#issuecomment-237155611), it refers for some failure in `instructorCommentsPageEditFrc.html` which indeed has been fixed, but the issues for `instructorCommentsPageAddFrc.html` and `instructorCommentsPageDeleteFrc.html` still persist.

**Comment by belyabl9** [comment]
@wkurniawan07 , I'll have a look on it today. As I remember, I checked it previously and it appeared only via running test suite. In a debug mode it worked as expected

**Comment by belyabl9** [comment]
@wkurniawan07 , as I previously told, when I ran all the tests, I got the previous result. When I run single test (InstructorCommentsPageUiTest), I get the diff as in the last commit. The most interesting thing for me is that Travis passed both variants.
So, please, have a look at the last commit

**Comment by belyabl9** [comment]
Ready to review

**Comment by wkurniawan07** [comment]
@belyabl9 I've somewhat isolated the cause of the strange problem: it seems like the test driver cannot access the rich text editor in the comments page if the browser is not active (e.g in background). Can you try to fix it?

**Comment by wkurniawan07** [comment]
Hi @belyabl9, any updates on this PR?

**Comment by belyabl9** [comment]
Guys, sorry for delay. My ISP stopped providing its services, so currently I'm looking for a new one.
I'll return as soon as I restore stable internet connection

**Comment by wkurniawan07** [comment]
Hi @belyabl9, any updates on this PR and/or your ISP status?

**Comment by belyabl9** [comment]
@wkurniawan07 , new provider promised to connect me this week, I'll notify ASAP

**Comment by wkurniawan07** [comment]
@belyabl9 thanks for the update. While waiting, can you try your best to work on this offline? TEAMMATES has been configured to be able to be developed fully offline, save for one UI test which I don't think is affected by this PR.

**Comment by belyabl9** [comment]
@wkurniawan07 , I'm an internet user again :) I've rebased the code, fixed some found bugs.
Could you, please, review it and point me out on the issues which still hold this code from being merged

Ready to review

**Comment by wkurniawan07** [comment]
@belyabl9 sadly the problem that I pointed out before still persists: if `InstructorCommentsPageUiTest` is running while the Firefox browser is on the background (meaning you work on something else on the foreground when the test is running), it will fail. Seems like the Selenium driver cannot access the rich text editor, although it might also be a javascript issue. Can you try your best to fix it?

**Comment by belyabl9** [comment]
@wkurniawan07 , I can't reproduce the described issue. Here are my steps:
1. Run InstructorCommentsPageUiTest (firefox)
2. Surf the web via Chrome on the foreground
3. Check the result after some time
Is it the same you are doing to reproduce it ? Could you add more details, because I have successful result regardless of firefox is in the foreground or background

**Comment by damithc** [comment]
Perhaps the issue is OS specific?

**Comment by belyabl9** [comment]
ping

**Comment by wkurniawan07** [comment]
@belyabl9 sorry for the delay, because this is one of the PRs that will take some time to review. I will now be continuing with the review.
I shall try running on my Windows machine. On my Mac and Travis CI which is running Linux the failure is confirmed, and I'd very much prefer if it can be fixed because it may be affecting the CI run in the long run.

Also, there are merge conflicts that needs resolve.

File: src/test/resources/pages/allJsUnitTests.html [hunk 20]
```diff
@@ -8,7 +8,9 @@
   <script type="text/javascript" src="../../javascript/lib/blanket.min.js"></script>
   <script type="text/javascript" src="../../../main/webapp/js/lib/jquery.min.js"></script>
   <script type="text/javascript" src="../../../main/webapp/js/lib/bootstrap.min.js"></script>
-  <script type="text/javascript" src="../../../main/webapp/js/common.js" data-cover></script>  
+  <script type="text/javascript" src="../../../main/webapp/js/richTextEditor.js" data-cover></script>
+  <script type="text/javascript" src="../../../main/webapp/js/common.js" data-cover></script>
+  <script type="text/javascript" src="../../../main/webapp/js/lib/tinymce.min.js"></script>
```
Comment (wkurniawan07) [inline src/test/resources/pages/allJsUnitTests.html]: This should come before `richTextEditor.js`.

**Comment by wkurniawan07** [comment]
It's reproducible in my Windows machine so it doesn't seem to be OS-specific, but we may take a look on it on another issue.

**Comment by belyabl9** [comment]
@wkurniawan07 , ok. I'll do it on Friday
